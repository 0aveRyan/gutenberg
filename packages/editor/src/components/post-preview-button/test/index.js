/**
 * External dependencies
 */
import { shallow } from 'enzyme';

/**
 * Internal dependencies
 */
import { PostPreviewButton } from '../';

describe( 'PostPreviewButton', () => {
	describe( 'setPreviewWindowLink()', () => {
		it( 'should do nothing if there is no preview window', () => {
			const url = 'https://wordpress.org';
			const setter = jest.fn();
			const wrapper = shallow( <PostPreviewButton /> );

			wrapper.instance().setPreviewWindowLink( url );

			expect( setter ).not.toHaveBeenCalled();
		} );

		it( 'set preview window location to url', () => {
			const url = 'https://wordpress.org';
			const setter = jest.fn();
			const wrapper = shallow( <PostPreviewButton /> );
			wrapper.instance().previewWindow = {
				get location() {
					return {
						href: 'about:blank',
					};
				},
				set location( value ) {
					setter( value );
				},
			};

			wrapper.instance().setPreviewWindowLink( url );

			expect( setter ).toHaveBeenCalledWith( url );
		} );
	} );

	describe( 'getWindowTarget()', () => {
		it( 'returns a string unique to the post id', () => {
			const instance = new PostPreviewButton( {
				postId: 1,
			} );

			expect( instance.getWindowTarget() ).toBe( 'wp-preview-1' );
		} );
	} );

	describe( 'componentDidUpdate()', () => {
		it( 'should change popup location if preview link is available', () => {
			const wrapper = shallow(
				<PostPreviewButton
					postId={ 1 }
					currentPostLink="https://wordpress.org/?p=1"
					isSaveable
					modified="2017-08-03T15:05:50" />
			);

			const previewWindow = { location: {} };

			wrapper.instance().previewWindow = previewWindow;

			wrapper.setProps( { previewLink: 'https://wordpress.org/?p=1' } );

			expect( previewWindow.location ).toBe( 'https://wordpress.org/?p=1' );
		} );
	} );

	describe( 'openPreviewOverlay()', () => {
		it( 'behaves like a regular link if not autosaveable', () => {
			const autosave = jest.fn();

			const wrapper = shallow(
				<PostPreviewButton
					postId={ 1 }
					autosave={ autosave }
				/>
			);

			wrapper.instance().openPreviewOverlay();

			expect( autosave ).not.toHaveBeenCalled();
		} );

		it( 'autosaves the post if autosaveable', () => {
			const autosave = jest.fn();

			const wrapper = shallow(
				<PostPreviewButton
					postId={ 1 }
					autosave={ autosave }
					isAutosaveable
				/>
			);

			wrapper.instance().openPreviewOverlay();

			expect( autosave ).toHaveBeenCalled();
		} );
	} );

	describe( 'openPreviewInNewTab()', () => {
		let windowOpen;
		beforeEach( () => {
			windowOpen = window.open;
		} );
		afterEach( () => {
			window.open = windowOpen;
		} );

		it( 'opens preview in new tab', () => {
			const preventDefault = jest.fn();
			const setLocation = jest.fn();
			window.open = jest.fn( () => ( {
				focus: jest.fn(),
				set location( url ) {
					setLocation( url );
				},
			} ) );

			const wrapper = shallow(
				<PostPreviewButton
					postId={ 1 }
				/>
			);

			wrapper.instance().openPreviewInNewTab( {
				preventDefault,
				target: { href: 'https://wordpress.org/?p=1' },
			} );

			expect( preventDefault ).toHaveBeenCalled();
			expect( window.open ).toHaveBeenCalledWith( '', '_blank' );
			expect( wrapper.instance().previewWindow.focus ).toHaveBeenCalled();
			expect( setLocation ).toHaveBeenCalledWith( 'https://wordpress.org/?p=1' );
		} );
	} );

	describe( 'render()', () => {
		const previewLink = 'https://wordpress.org/?p=1&preview=true';
		const currentPostLink = 'https://wordpress.org/?p=1';
		it( 'should render overlay when state is open', () => {
			const wrapper = shallow(
				<PostPreviewButton
					postId={ 1 }
					currentPostLink={ currentPostLink }
				/>
			).setState( { isPreviewOpen: true } );

			expect( wrapper ).toMatchSnapshot();
		} );

		it( 'should render previewLink if provided', () => {
			const wrapper = shallow(
				<PostPreviewButton
					postId={ 1 }
					isSaveable
					previewLink={ previewLink }
					currentPostLink={ currentPostLink }
				/>
			).setState( { isPreviewOpen: true } );

			const frameSrc = wrapper.find( '.editor-block-preview__frame' ).prop( 'src' );
			expect( frameSrc ).toEqual( previewLink );
		} );

		it( 'should render currentPostLink otherwise', () => {
			const wrapper = shallow(
				<PostPreviewButton
					postId={ 1 }
					isSaveable
					currentPostLink={ currentPostLink }
				/>
			).setState( { isPreviewOpen: true } );

			const frameSrc = wrapper.find( '.editor-block-preview__frame' ).prop( 'src' );
			expect( frameSrc ).toEqual( currentPostLink );
		} );

		it( 'should be disabled if post is not saveable', () => {
			const wrapper = shallow(
				<PostPreviewButton
					postId={ 1 }
					currentPostLink="https://wordpress.org/?p=1"
				/>
			).find( '.editor-post-preview' );

			expect( wrapper.prop( 'disabled' ) ).toBe( true );
		} );
	} );
} );
