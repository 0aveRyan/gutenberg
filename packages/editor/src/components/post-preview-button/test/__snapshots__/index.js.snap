// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostPreviewButton render() should render overlay when state is open 1`] = `
<Fragment>
  <ForwardRef(Button)
    className="editor-post-preview"
    disabled={true}
    isLarge={true}
    onClick={[Function]}
  >
    Preview
    <WithSelect(WithDispatch(DotTip))
      tipId="core/editor.preview"
    >
      Click “Preview” to load a preview of this page, so you can make sure you’re happy with your blocks.
    </WithSelect(WithDispatch(DotTip))>
  </ForwardRef(Button)>
  <WithInstanceId(Modal)
    className="editor-block-preview"
    onRequestClose={[Function]}
    shouldCloseOnClickOutside={false}
    title="Preview"
  >
    <div
      className="editor-block-preview__controls"
    >
      <ForwardRef(IconButton)
        icon={
          <SVG
            height="24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <Path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <Path
              d="M21 2H3c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h7v2H8v2h8v-2h-2v-2h7c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm0 14H3V4h18v12z"
            />
          </SVG>
        }
        label="Preview desktop screen"
        onClick={[Function]}
      />
      <ForwardRef(IconButton)
        icon={
          <SVG
            height="24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <Path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <Path
              d="M21 4H3c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h18c1.1 0 1.99-.9 1.99-2L23 6c0-1.1-.9-2-2-2zm-2 14H5V6h14v12z"
            />
          </SVG>
        }
        label="Preview tablet screen"
        onClick={[Function]}
      />
      <ForwardRef(IconButton)
        icon={
          <SVG
            height="24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <Path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <Path
              d="M17 1.01L7 1c-1.1 0-2 .9-2 2v18c0 1.1.9 2 2 2h10c1.1 0 2-.9 2-2V3c0-1.1-.9-1.99-2-1.99zM17 19H7V5h10v14z"
            />
          </SVG>
        }
        label="Preview phone screen"
        onClick={[Function]}
      />
      <ForwardRef(Button)
        className="editor-block-preview__new-tab"
        href="https://wordpress.org/?p=1"
        isLarge={true}
        onClick={[Function]}
        target="wp-preview-1"
      >
        Open preview in new tab
      </ForwardRef(Button)>
    </div>
    <div
      className="editor-block-preview__frame-container"
    >
      <iframe
        className="editor-block-preview__frame"
        src="https://wordpress.org/?p=1"
        style={
          Object {
            "height": "800px",
            "width": "1200px",
          }
        }
        title="Responsive preview frame"
      />
    </div>
  </WithInstanceId(Modal)>
</Fragment>
`;
